plugins/modules/ec2_asg_tag.py

```diff
@@ -0,0 +1,237 @@
+      - Optional key is I(propagate_at_launch), which defaults to true.
+    type: list
+    elements: dict
+  purge_tags:
+    description:
+      - Whether unspecified tags should be removed from the resource.
+      - Note that when combined with I(state=absent), specified tags with non-matching values are not purged.
+    type: bool
+    default: false
+extends_documentation_fragment:
+- amazon.aws.aws
+- amazon.aws.ec2
+'''
+
+EXAMPLES = r'''
+- name: Ensure tags are present on an ASG
+  community.aws.ec2_asg_tag:
+    name: my-auto-scaling-group
+    state: present
+    tags:
+      - environment: production
+        propagate_at_launch: true
+      - role: webserver
+        propagate_at_launch: true
+
+- name: Ensure tag is absent on an ASG
+  community.aws.ec2_asg_tag:
+    name: my-auto-scaling-group
+    state: absent
+    tags:
+      - environment: development
+
+- name: Remove all tags except Name from an ASG
+  community.aws.ec2_asg_tag:
+    name: my-auto-scaling-group
+    state: absent
+    tags:
+      - Name: ''
+    purge_tags: true
+'''
+
+RETURN = r'''
+---
+tags:
+  description: A list containing the tags on the resource
+  returned: always
+  type: list
+  sample: [
+      {
+          "key": "Name",
+          "value": "public-webapp-production-1",
+          "resource_id": "public-webapp-production-1",
+          "resource_type": "auto-scaling-group",
+          "propagate_at_launch": "true"
+      },
+      {
+          "key": "env",
+          "value": "production",
+          "resource_id": "public-webapp-production-1",
+          "resource_type": "auto-scaling-group",
+          "propagate_at_launch": "true"
+      }
+  ]
+added_tags:
+  description: A list of tags that were added to the ASG
+  returned: If tags were added
+  type: list
+removed_tags:
+  description: A list of tags that were removed from the ASG
+  returned: If tags were removed
+  type: list
+'''
+
+try:
+    from botocore.exceptions import BotoCoreError, ClientError
+except ImportError:
+    pass  # Handled by AnsibleAWSModule
+
+from ansible.module_utils._text import to_native
+
+from ansible_collections.amazon.aws.plugins.module_utils.core import AnsibleAWSModule
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import AWSRetry
+
+
+def to_boto3_tag_list(tags, group_name):
+    tag_list = []
+    for tag in tags:
+        for k, v in tag.items():
+            if k == 'propagate_at_launch':
+                continue
+            tag_list.append(dict(Key=k,
+                                 Value=to_native(v) if v else v,
+                                 PropagateAtLaunch=bool(tag.get('propagate_at_launch', True)),
+                                 ResourceType='auto-scaling-group',
+                                 ResourceId=group_name))
+    return tag_list
+
+
+def compare_asg_tags(current_tags_dict, new_tags_dict, purge_tags=True):
```

Comment:
Could you please import and use the `compare_aws_tags` function from the module utility?
https://github.com/ansible-collections/amazon.aws/blob/main/plugins/module_utils/ec2.py#L783

Reply:
@jillr this and the other requested changes weren't possible based on my testing. 

ASG tags aren't just K/V pairs (as expected by `compare_aws_tags`, `boto3_tag_list_to_ansible_dict` and `ansible_dict_to_boto3_tag_list`), they're a list of dicts instead. The ramification here is that if `propagate_at_launch` is changed but the K/V doesn't, the module won't detect any tag changes which I believe is not the desired behaviour.

Another issue I ran into was that the `to_text` and `to_native` functions did some odd things, particularly around `None` values. Hence why `compare_asg_tags` doesn't use them like `compare_aws_tags`.
From memory you can't roundtrip `None` like `to_native(to_text(None))`... or maybe it was the other way around? 🤔